packages/types/src/type/ExtrinsicEra.ts

```diff
@@ -2,33 +2,166 @@
 // This software may be modified and distributed under the terms
 // of the Apache-2.0 license. See the LICENSE file for details.
 
+import EnumType from '../codec/EnumType';
+import {isHex, isU8a, hexToU8a, isObject, assert, u8aToBn} from '@polkadot/util';
+import Tuple from '../codec/Tuple';
+import U64 from '../primitive/U64';
+import U8 from '../primitive/U8';
+import U16 from '../primitive/U16';
 import { AnyU8a } from '../types';
+import BlockNumber from "@polkadot/types/type/BlockNumber";
+import U8a from "@polkadot/types/codec/U8a";
 
-import { u8aToU8a } from '@polkadot/util';
+interface EraMethod {
+  current: BlockNumber;
+  period: BlockNumber;
+}
+
+export default class ExtrinsicEra extends EnumType<ImmortalEra | MortalEra> {
+  constructor (value?: any, index?: number) {
+    super({ ImmortalEra, MortalEra }, value, index);
+  }
+
+  /**
+   * @description Returns the item as a [[ImmortalEra]]
+   */
+  get asImmortalEra (): ImmortalEra {
+    assert(this.isImmortalEra, `Cannot convert '${this.type}' via asImmortalEra`);
+    return this.value as ImmortalEra;
+  }
+
+  /**
+   * @description Returns the item as a [[MortalEra]]
+   */
+  get asMortalEra (): MortalEra {
+    assert(this.isMortalEra, `Cannot convert '${this.type}' via asMortalEra`);
+    return this.value as MortalEra;
+  }
+
+  /**
+   * @description `true` if Immortal
+   */
+  get isImmortalEra (): boolean {
+    return this.type === 'ImmortalEra';
+  }
+
+  /**
+   * @description `true` if Mortal
+   */
+  get isMortalEra (): boolean {
+    return this.type === 'MortalEra';
+  }
 
-import U8a from '../codec/U8a';
+  /**
+   * @description Encodes the value as a Uint8Array as per the parity-codec specifications
+   * @param isBare true when the value has none of the type-specific prefixes (internal)
+   */
+  toU8a (isBare?: boolean): Uint8Array {
+    if (this.index === 0 ) {
+      return super.toU8a()
+    } else {
+      return (this.asMortalEra as MortalEra).toU8a(isBare);
+    }
+  }
+}
+
+export class ImmortalEra extends U8a {
+  constructor(value?: AnyU8a) {
+    super(value);
+  }
+}
+
+export type MortalEraValue = [U8, U8];
 
 /**
- * @name ExtrinsicEra
+ * @name MortalEra
  * @description
- * The era for an extrinsic, indicating either a mortal or immortal extrinsic
+ * The MortalEra for an extrinsic, indicating period and phase
  */
-export default class ExtrinsicEra extends U8a {
-  constructor (value?: AnyU8a) {
-    super(
-      ExtrinsicEra.decodeExtrinsicEra(value)
-    );
+export class MortalEra extends Tuple {
+  constructor (value?: any) {
+    super({
+      period : U8, phase : U8
+    }, MortalEra.decodeMortalEra(value));
+  }
+
+  private static decodeMortalEra (value: EraMethod | Uint8Array | string): MortalEraValue {
```

Comment:
if value is object it actually means two possibilities, one is `{current, period}`, the other is `{period, phase}`.
we should avoid to make it ambiguous.